### ISSUE TITLE
Pre-training Using Spectral Clustering 

### ISSUE THREAD

**Issue by KisaragiDev** [kisaragiRY/decoding-DG#47]
# Spectral clustering for choosing data parameters
- Method: Spectral Clustering 
- Score: mean similarity of samples within the clusters
- Parameters:
   - `window size`: window size of Gaussian kernel for smoothing data
   - `K`: the threshold for the length of segments
   - `num_kernels`: number of kernels used in ROCKET transformation
- Criteria: the higher the mean similarity is of the clustering model, the better.
- Desired results: as `num_kernels` increases, the score also increases.
- working dir: `/work/notebooks/3-decoding-model/2-classification/3-time-series-classification/2-pre-training`

## Work Flow
- [x] research on how to use Spectral Clustering 
- [x] train script for testing on diff params of data -> `train.py`
   1. large scale of num_kernels and K 
   2. small scale
- [x] Analyze the training results -> `02-influence-of-data-param.ipynb`
   1. get the best param pair from the large scale
   2. examine the small scale results
- [x] Analyze the similarity between the clustering results and y(position) -> `03-similarity-between-clusters-positions.ipynb`
   * coverage rate, e.g probability of instances from the same position are in the same cluster
   * if the coverage rate is above 50% ?

**Comment by KisaragiDev** [comment]
# Usage of [Sklearn Spectral Clustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html#sklearn.cluster.SpectralClustering)
```python
from sklearn.cluster import SpectralClustering
clustering = SpectralClustering(
     n_clusters=4,
     assign_labels='discretize',
     random_state=0)
clustering.fit(X_train)

# affinity matrix(similarity matrix after fitting)
clustering.affinity_matrix_
```

**Comment by KisaragiDev** [comment]
# To evaluate the training results
- [x] calculate the inter clusters similarity based on affinity matrix -> `cal_inter_cluster_similarity`
   1. get the indices of samples from the cluster 1 and cluster 2
   2. extract the values of those samples from affinity matrix 
   3. calculate the mean value of those values
   4. repeat above for all the other clusters combinations
   5. calculate the mean of results from different clusters combinations

- [x] calculate the within clusters similarity based on affinity matrix -> `cal_within_cluster_similarity`

## Similarly for the K-means results
- [x] inter cluster distance -> `cal_inter_clusters_distance`
- [x] within cluster distance -> `cal_within_clusters_distance`

# Next 
Manipulating `cal_inter_cluster_similarity` to get a desired results

**Comment by KisaragiDev** [comment]
# Calculating the inter cluster similarity

## Constructing Affinity Matrix
> using a Gaussian Kernel based on the Euclidean distance d(X, X)
---- sklearn

similarity between $X_1$ and $X_2$:  `np.exp(-gamma * d(X1,X2) ** 2)`

## Assigning Labels
a column-pivoted QR factorization
[Simple, direct and efficient multi-way spectral clustering](https://academic.oup.com/imaiai/article/8/1/181/5045955?login=true)

## Final Result
```python
def cal_inter_cluster_similarity(affinity_matrix, clusters_labels):
    """Calculate the inter clusters similarity based on affinity matrix.
    """
    clusters= np.unique(clusters_labels)
    n_clusters = len(clusters)
    mean_inter_clusters = np.zeros(int(np.math.factorial(n_clusters)/
                                   (np.math.factorial(n_clusters-2)*np.math.factorial(2))))
    # for cluster_id, cluster in enumerate(clusters):
    for combo_id, combo in enumerate(combinations(range(n_clusters), 2)):
        sample_ids_1 = np.where(clusters_labels==combo[0])[0]
        sample_ids_2 = np.where(clusters_labels==combo[1])[0]
        filtered = affinity_matrix[sample_ids_1][:, sample_ids_2]
        # filtered = [affinity_matrix.todok()[combo_sample] for combo_sample in product(sample_ids_1, sample_ids_2)]
        mean_inter_clusters[combo_id] = np.max(filtered)
    return np.median(mean_inter_clusters)
```

**Comment by KisaragiDev** [comment]
# Observation
1. The affinity matrix is also based on the euclidean distance, so there is no distinct different between the K-means and Spectral clustering.
Therefore, **we should just focus on the spectral clustering** since it outputs the distance(similarity) between each instance.
2. the inter-clusters sim drops roughly after $2^4$, and the change starting to fluctuate less around $2^8$
**A more detailed check-up should be done around $2^8$**
3. Though the number of clusters were set to 4, **the relationship between the 4 clusters and the 4 actual position labels remains unknown.**

**Comment by KisaragiDev** [comment]
# Coverage Rate
1. Each label (internal state) predicted from K-means model corresponds to several positions.
2. Take the position with highest correspondence as the predicted position.
3. The accuracy is called coverage rate here

## Observation
- With K-means, the coverage rate for both KO and WT mice are around 42%
- With Spectral Clustering, KO is around 35% and WT is around 43%
